### ISSUE TITLE
Error when merging tickets in Zammad, if the default name for the status "merged" was changed.

### ISSUE THREAD

**Issue by MarkusHeld** [zammad/zammad#5191]
### Used Zammad Version

6.3

### Environment

- Installation method: any
- Operating system: any
- Database + version:  any
- Elasticsearch version: any

### Actual behaviour

If the (default) name for the status “merged” was changed and then you want to "merge" a ticket in Zammad you get following error.

```
I, [2024-05-17T07:22:08.690831#964427-10000360]  INFO -- : Started PUT "/api/v1/ticket_merge/34945/202405012700024" for 80.151.107.194 at 2024-05-17 07:22:08 +0000
I, [2024-05-17T07:22:08.695283#964427-10000360]  INFO -- : Processing by TicketsController#ticket_merge as JSON
I, [2024-05-17T07:22:08.695343#964427-10000360]  INFO -- :   Parameters: {"source_ticket_id"=>"34945", "target_ticket_number"=>"202405012700024"}
E, [2024-05-17T07:22:08.848789#964427-10000360] ERROR -- : undefined method `id' for nil:NilClass (NoMethodError)
app/models/ticket.rb:413:in `block in merge_to'
app/models/transaction.rb:17:in `block in execute'
app/models/transaction.rb:14:in `execute'
app/models/transaction.rb:26:in `execute'
app/models/ticket.rb:337:in `merge_to'
app/services/service/ticket/merge.rb:9:in `execute'
app/controllers/tickets_controller.rb:422:in `ticket_merge'
app/controllers/application_controller/has_download.rb:17:in `block (4 levels) in <module:HasDownload>'
app/controllers/application_controller/has_download.rb:16:in `block (3 levels) in <module:HasDownload>'
app/controllers/application_controller/has_download.rb:15:in `block (2 levels) in <module:HasDownload>'
app/controllers/application_controller/handles_transitions.rb:16:in `handle_transaction'
I, [2024-05-17T07:22:08.852493#964427-10000360]  INFO -- : Completed 500 Internal Server Error in 157ms (Views: 0.1ms | ActiveRecord: 28.1ms | Allocations: 41166)
```

# app/models/ticket.rb:413

self.state_id = Ticket::State.lookup(name: 'merged').id

### Expected behaviour

Zammad recognizes that the name for the status has changed and it is possible to merge tickets. 

### Steps to reproduce the behaviour

- Change the name of the status "merged" for example to "zusammengefasst.
- Try in Zammad to merge a ticket to another ticket.

### Support Ticket

Ticket#10155116, #10159302, #10160205, #10160857, #10167175

### I'm sure this is a bug and no feature request or a general question.

yes

**Comment by dominikklein** [comment]
I think this is not really related to the import any longer, because of the new admin interface to change the ticket states. So it should be the same problem, when you rename the default ticket state name from `merged` to something else inside the admin interface.

**Comment by fliebe92** [comment]
We have a lot of locations in the code where we're checking for `merged` state name. All of them might be broken when someone is changing the `merged` ticket state to something else.

It's not related to the OTRS import.

**Comment by MarkusHeld** [comment]
Ok, thanks for the note. I was also able to reproduce the error when I changed the name for “merged” in the object in Zammad.

I have now adjusted the issue.

**Comment by martini** [comment]
IMO the state `merge` should not be editable (there is also no usecase to rename it).

**Comment by mantas** [comment]
Or we could make sure it's always detected by state type name, not by state itself.

**Comment by martini** [comment]
> Or we could make sure it's always detected by state type name, not by state itself.

What if you have created two states with state type merged?

**Comment by mantas** [comment]
> > Or we could make sure it's always detected by state type name, not by state itself.
> 
> What if you have created two states with state type merged?

I agree this would be abusing the system. But in such case I think it'd be better to treat both states as merged-by-type. Rather than take one as merged-by-name and treat the other one as..... Hmmm.... No idea how to be honest.

**Comment by t-shehab** [comment]
If we're going to look for hardcoded `name: "merged"` in the code, then I don't see any way, other than disallowing the renaming of that state, to fix the problem.

Another Idea could be an additional attribute to the `Ticket::State` object. Something like `name_initial` maybe, which never changes. Then search for this attribute in the code, as opposed to searching for `name`.

**Comment by MrGeneration** [comment]
This occurred at least twice within the last 14 days in SaaS again. Please, we need a solution on this. I personally am voting for write protected as well, as the logic that use this are hard coded. Or make it dynamically look for the state type (which might introduce other issues...)

**Comment by mantas** [comment]
I'd like to propose following solution:

- Switch existing checks from merged state name to state type name. Looks like there're 2 occurrences in our code base.

- Block creation of additional merged states. Keep renaming functionality.

Proof of concept MR turned out pretty nice: https://git.zammad.com/zammad/zammad/-/merge_requests/8712

**Comment by dominikklein** [comment]
@YetAnotherGerrit Here we would need clarification if renaming the `merged` state is really a use case we want to solve in our product. The question would also be if we need to clarify this for other states.

**Comment by YetAnotherGerrit** [comment]
Can you help me understand, why we check for the state "merged" and not for the state type "merged"?

And can we prevent the user to have no state with the state type "merged"? So, if they try change the state type of the state "merged" and there is no other state with the state type "merged", to disallow that? E.g. disable the state type selector if this is the only state with the state type "merged".

**Comment by YetAnotherGerrit** [comment]
I checked your MR @mantas - you prevented to create/updated a Ticket State, so a **max. of 1** state with the state type "merged" is possible. But can you also enforce an **min. of 1**? E.g. to not be able to "remove" the "merged" state type from the last existing state with that state type.

Other than that, the MR looks fine.

**Comment by mantas** [comment]
@YetAnotherGerrit it's not allowed to remove a state (or priority) via UI. I doubt it's worth to add this limitation to command line. If someone is determined to shoot himself in a foot, he can go directly into database.

**Comment by mantas** [comment]
Note from live discussion with @YetAnotherGerrit . Technically it's not possible to remove the state. But it's still possible to change the type of it. Effectively having a system without a `merged` type state. Validation should be added to prevent changing of state type in such cases.

**Comment by dominikklein** [comment]
@YetAnotherGerrit I think the last decision was that it should still be possible to rename such "static" single states or? But it should not be possible to remove the last one of these states (this means no delete functionality and not changing the state type) and create additional ones.

**Comment by YetAnotherGerrit** [comment]
Correct. Mantas showed me a preview of that solution on the off-site.

**Comment by mantas** [comment]
So if I remember correctly we have 2 issues in my old branch:

- `merged` state cannot be removed, but type can be changed. So we should prevent changing a type in such a way that no `merged` states exist anymore.

- There're many places in the code base that detect `merged` state. All of them needs to replaced.

Is that correct?

**Comment by YetAnotherGerrit** [comment]
Is there a preview-app? Then I will have a look and see if I remember/find additional todos.

**Comment by mantas** [comment]
@YetAnotherGerrit feel free to start a review app in https://git.zammad.com/zammad/zammad/-/merge_requests/8712/pipelines when you're ready to test. I just rebased it, but if you'll test sometime later, rebasing and starting a fresh app may be better..

There're comments from Rolf on that MR too about complex (but invisible) changes needed to make it work.

**Comment by YetAnotherGerrit** [comment]
If we decide to go the route of making `merged` a read only state, is there a way to allow admins to fix their system, if they already renamed their `merged` state or created additional ones?

**Comment by mantas** [comment]
I could see two possible migration paths:

a) Rename oldest `merged` state to `merged`. Then change newer `merged` states to `closed` type.

b) Rename oldest `merged` state to `merged`. Then move all tickets from newer `merged` states to that oldest one and delete those states.

**Comment by dominikklein** [comment]
For me, option b) sounds more understandable.

**Comment by YetAnotherGerrit** [comment]
It does not sound appealing to me to delete something from a customer system. They created these states for a reason. It might be quirky and a bad idea what they did, but we should not destroy the information if a ticket has been "merged a" or "merged b". With a) we can preserve at least the information that there are two tickets with two different states in that example.

Another question if we go the "read only"-route: can we prevent changing the state via Console/API or will this be a front-end restriction only?

Actually, I don't like both approaches we talked so far. The reason: we have a third approach implemented already for "new" and "open" and that is, that we need to select the "default new" and "default open" state. Is there a reason, why we are not choosing the same solution for "merged"? This would be consistent.

**Comment by mantas** [comment]
> Another question if we go the "read only"-route: can we prevent changing the state via Console/API or will this be a front-end restriction only?

Yes, it'd be on console/api level.

> Actually, I don't like both approaches we talked so far. The reason: we have a third approach implemented already for "new" and "open" and that is, that we need to select the "default new" and "default open" state. Is there a reason, why we are not choosing the same solution for "merged"? This would be consistent.

The main problem IMO is that there're multiple places in the codebase that just picks state with `name: merged`. And tracking down all of them would take time. I've a feeling some of those parts don't have elaborate test coverage too. It sounds like a rather risky endeavour.

But I agree that `default merge` or something like that would technically be the best solution. We may need to come up with some creative solutions in some places though. For example, reporting. Maybe count any merge-like states. Not sure how new/open operates there right now.

**Comment by YetAnotherGerrit** [comment]
Okay, thank you for the insights. We will do the creative solution for now that will fulfill the needs of all stakeholders:
- Make the `merged` state read-only.
- Prevent the user from creating additional `merged` states.
- Choose migration path a) Rename oldest `merged` state to `merged`. Then change newer `merged` states to `closed` type.

As this might be a breaking change and the urgency is minor, we will announce it in the next release and ship it with the release after that.

But if we ever touch the state thingy again, we should keep the `default merge` solution in mind.

**Comment by mantas** [comment]
Sounds good to me

**Comment by dominikklein** [comment]
I think the merge feature is nothing where it is supposed to have multiple merged states. You cannot choose them because it will never be possible to use different merge states.

But will we add the read-only functionality now? Otherwise, we run again in support tickets? It's one of the issues with more ticket numbers behind.
What kind of benefit would people have from announcing the breaking change? Can say solve anything her eon they own before this migration is incoming?

**Comment by YetAnotherGerrit** [comment]
> I think the merge feature is nothing where it is supposed to have multiple merged states. You cannot choose them because it will never be possible to use different merge states.

Yeah, that's why we will make it read only. People will not be able to change the one `merged` state and will not be able to create additional ones.

> But will we add the read-only functionality now?

Yes.

> Otherwise, we run again in support tickets?

Yes.

> What kind of benefit would people have from announcing the breaking change?

If they have more than one `merged` state, we will give them the opportunity to sort that out themselves. Instead of forcing them to `closed` state types. Also, we will provide them with the opportunity to "save" their current name for `merged` into the translations.

> Can say solve anything her eon they own before this migration is incoming?

Yes, affected admins can currently solve this issue on their own, so there is no need to rush things. The changes we'll implement are mainly to prevent future admins to run into this situation.
